program/src/processor/set_authority.rs

```diff
@@ -13,9 +19,85 @@ use crate::instruction::{
 ///   1. `[s]` config_authority
 ///   2. `[]` new_authority
 pub fn process_set_authority(
-    _program_id: &Pubkey,
-    _ctx: Context<SetAuthorityAccounts>,
-    _authority_type: AuthorityType,
+    program_id: &Pubkey,
+    ctx: Context<SetAuthorityAccounts>,
+    authority_type: AuthorityType,
 ) -> ProgramResult {
+    // Accounts validation.
+
+    // 1. authority
+    // - must be a signer
+    // - must match the authority in the account (checked in the match statement below)
+
+    require!(
+        ctx.accounts.authority.is_signer,
+        ProgramError::MissingRequiredSignature,
+        "authority"
+    );
+
+    // 2. account
+    // - owner must the stake program
+    // - must be initialized
+    // - must have an authority set
+
+    require!(
+        ctx.accounts.account.owner == program_id,
+        ProgramError::InvalidAccountOwner,
+        "account"
+    );
+
+    let data = &mut ctx.accounts.account.try_borrow_mut_data()?;
+
+    match authority_type {
+        AuthorityType::Config | AuthorityType::Slash => {
+            let config = bytemuck::try_from_bytes_mut::<Config>(data)
+                .map_err(|_error| ProgramError::InvalidAccountData)?;
+
+            require!(
+                config.is_initialized(),
+                ProgramError::UninitializedAccount,
+                "config"
+            );
+
+            // Asserts that the authority is set - only updates the authority jf there is
+            // one set and it matches the signing authority.
+            //
+            // TODO: do we fail if the authority is not set?
+
+            let current_authority = config.authority.into();
```

Comment:
Sorry, this wasn't clear, but I was thinking this would operate similarly to the different authorities in an SPL token mint, where each authority has control over itself and that's it.

So the config authority can set another config authority, and the slash authority can set another slash authority. This way we can eventually delegate those authorities to different programs. What do you think?

Reply:
Got it, makes sense. I will change the implementation to reflect this.